File: src/migrations.ts
import { copyFileSync, existsSync } from 'node:fs';
import type { Database } from 'bun:sqlite';

/** 스키마 마이그레이션 하나 — 같은 트랜잭션 안에서 실행된다. */
export type Migration = (db: Database) => void;

/**
 * 마이그레이션 1: todo/note 에 보드별 순번(number)을 부여한다.
 *
 * 랜덤 id 는 PK 로 그대로 두고 표시·참조용 번호만 더한다. 기존 행에는 보드별로
 * created_at 순(동률이면 id 순 — 같은 밀리초 생성의 결정성)으로 1부터 소급 부여한다.
 */
const addNumbers: Migration = (db) => {
  db.run('ALTER TABLE todos ADD COLUMN number INTEGER');
  db.run('ALTER TABLE notes ADD COLUMN number INTEGER');

  for (const table of ['todos', 'notes'] as const) {
    const rows = db
      .query<{ id: string; board_id: string | null }, []>(
        `SELECT id, board_id FROM ${table} ORDER BY board_id, created_at ASC, id ASC`,
      )
      .all();
    const counters = new Map<string, number>();
    const update = db.query(`UPDATE ${table} SET number = ? WHERE id = ?`);
    for (const row of rows) {
      const key = row.board_id ?? '';
      const next = (counters.get(key) ?? 0) + 1;
      counters.set(key, next);
      update.run(next, row.id);
    }
  }

  db.run('CREATE UNIQUE INDEX idx_todos_number ON todos(board_id, number)');
  db.run('CREATE UNIQUE INDEX idx_notes_number ON notes(board_id, number)');
  // notes.board_id 는 nullable 이고 SQLite 는 유니크 인덱스에서 NULL 을 서로 다른 값으로
  // 취급한다 — 글로벌 메모끼리의 유일성은 부분 인덱스로 따로 건다.
  db.run('CREATE UNIQUE INDEX idx_notes_number_global ON notes(number) WHERE board_id IS NULL');
};

/**
 * 마이그레이션 2: 핸드오프 큐 테이블.
 *
 * 신규 DB 는 `SCHEMA` 로 이 테이블을 갖고 태어나므로 `IF NOT EXISTS` 가드가 필요하다 —
 * 마이그레이션은 신규/기존 DB 양쪽에서 실행된다.
 */
const addHandoffs: Migration = (db) => {
  db.run(`
    CREATE TABLE IF NOT EXISTS handoffs (
      id            TEXT PRIMARY KEY,
      todo_id       TEXT NOT NULL REFERENCES todos(id),
      session_id    TEXT NOT NULL,
      session_name  TEXT,
      session_cwd   TEXT,
      note          TEXT NOT NULL DEFAULT '',
      actor         TEXT NOT NULL,
      status        TEXT NOT NULL CHECK (status IN ('pending','delivered','cancelled')),
      created_at    TEXT NOT NULL,
      delivered_at  TEXT,
      delivered_via TEXT
    )
  `);
  const columns = new Set(
    db
      .query<{ name: string }, []>('PRAGMA table_info(handoffs)')
      .all()
      .map((row) => row.name),
  );
  if (columns.has('session_id') && columns.has('status') && columns.has('created_at')) {
    db.run(
      'CREATE INDEX IF NOT EXISTS idx_handoffs_session ON handoffs(session_id, status, created_at)',
    );
  }
  if (columns.has('todo_id') && columns.has('status')) {
    db.run('CREATE INDEX IF NOT EXISTS idx_handoffs_todo ON handoffs(todo_id, status)');
  }
};

/** 적용 순서 = 배열 순서. 인덱스+1 이 곧 user_version. 기존 항목은 절대 수정하지 않는다. */
export const MIGRATIONS: Migration[] = [addNumbers, addHandoffs];

export interface RunMigrationsOptions {
  /** 테스트에서 목록을 주입한다. 기본은 MIGRATIONS. */
  migrations?: Migration[];
  /**
   * 적용 전 DB 를 복사해 둘 경로. 생략하면 `${dbPath}.bak-v<현재 user_version>` 로
   * 자동 계산한다 — 실제로 몇 버전에서 마이그레이션을 시작했는지가 파일명에 남아야
   * 여러 번 재기동을 거친 DB 의 백업 이력을 구분할 수 있다("v0" 로 고정하면 세 번째
   * 마이그레이션에서 만든 백업도 전부 v0 로 보여 오해를 부른다).
   */
  backupPath?: string;
  /**
   * 백업 원본 경로. dbPath 가 없거나(:memory: 등) 파일이 없거나, todos/notes 에 아직
   * 아무 행도 없는 신규 DB(백업할 내용이 없음)면 백업을 생략한다.
   */
  dbPath?: string;
}

/**
 * todos/notes 에 백업할 만한 데이터가 있는지 본다.
 *
 * 두 테이블 다 비어 있으면(막 만든 신규 DB) 백업이 무의미하다 — 신규 설치·임시 디렉터리
 * 테스트마다 `*.bak-v0` 잔재가 남는 걸 막는다. 테이블 자체가 없는 등 판단할 수 없는 경우는
 * 보수적으로 true(백업함) 를 반환한다.
 */
function hasDataWorthBackingUp(db: Database): boolean {
  try {
    const todos = db.query<{ n: number }, []>('SELECT COUNT(*) AS n FROM todos').get()?.n ?? 0;
    if (todos > 0) {
      return true;
    }
    const notes = db.query<{ n: number }, []>('SELECT COUNT(*) AS n FROM notes').get()?.n ?? 0;
    return notes > 0;
  } catch {
    return true;
  }
}

/**
 * WAL 을 메인 파일로 체크포인트한 뒤 파일 복사로 백업한다.
 *
 * `journal_mode = WAL` 에서는 마지막 체크포인트 이후의 커밋이 `-wal` 사이드카에만
 * 있고 메인 db 파일에는 없을 수 있다 — `copyFileSync` 로 메인 파일만 복사하면 그
 * 트랜잭션들이 통째로 빠진(심하면 스키마조차 없는) 백업이 만들어진다. 매 기동마다
 * `PRAGMA wal_checkpoint(TRUNCATE)` 로 WAL 을 메인 파일에 흡수시키고 `-wal` 을
 * 비운 뒤 복사한다 — `VACUUM INTO` 보다 저렴하고(테이블 전체 재작성이 없다), 이미
 * 열려 있는 handle 로 바로 실행할 수 있어 별도 연결이 필요 없다.
 *
 * 복사 자체가 실패하면(디스크 풀·권한) 예외를 던지지 않고 경고만 남긴 채 진행한다.
 * 백업 실패로 데몬 기동 전체가 막히면 "백업 없이 보드가 뜨는 것"보다 나쁜 결과를
 * 만든다 — 이 레포의 다른 fail-open 판단들(예: 버전 재기동 실패 시 구버전 유지)과
 * 같은 원칙이다. 정말 디스크가 가득 찼다면 뒤따르는 마이그레이션 자체(ALTER TABLE 등)도
 * 실패해 자연히 표면화된다.
 */
function backupDatabase(db: Database, dbPath: string, backupPath: string): void {
  try {
    db.run('PRAGMA wal_checkpoint(TRUNCATE)');
  } catch {
    // 체크포인트가 실패해도 최선을 다해 복사는 시도한다 — WAL 이 반영 안 된 백업이라도
    // 백업이 아예 없는 것보다는 낫다.
  }
  try {
    copyFileSync(dbPath, backupPath);
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    console.error(
      `rocky-todo: pre-migration backup failed (${backupPath}) — 백업 없이 진행: ${message}`,
    );
  }
}

/**
 * user_version 보다 뒤에 있는 마이그레이션만 순서대로 적용한다.
 *
 * 각 마이그레이션은 트랜잭션 안에서 실행되며, 던지면 롤백하고 user_version 도 올리지
 * 않는다 — 실패한 마이그레이션은 다음 기동에서 다시 시도된다. user_version 갱신은 같은
 * 트랜잭션 안에서 스키마 변경과 함께 커밋된다 — COMMIT 뒤 별도로 쓰면 그 사이 프로세스가
 * 죽었을 때 스키마는 적용됐는데 user_version 은 0 인 상태가 남아, 다음 기동에서 같은
 * 마이그레이션이 재실행되며 (예: ALTER TABLE 의 duplicate column) 영구히 기동 불가에 빠진다.
 * @returns 적용 후 최종 user_version.
 */
export function runMigrations(db: Database, options: RunMigrationsOptions = {}): number {
  const migrations = options.migrations ?? MIGRATIONS;
  const current =
    db.query<{ user_version: number }, []>('PRAGMA user_version').get()?.user_version ?? 0;
  if (current >= migrations.length) {
    return current;
  }

  if (options.dbPath && existsSync(options.dbPath) && hasDataWorthBackingUp(db)) {
    const backupPath = options.backupPath ?? `${options.dbPath}.bak-v${current}`;
    backupDatabase(db, options.dbPath, backupPath);
  }

  let version = current;
  for (let i = current; i < migrations.length; i++) {
    const migration = migrations[i];
    if (!migration) {
      continue;
    }
    db.run('BEGIN');
    try {
      migration(db);
      version = i + 1;
      // PRAGMA 는 바인딩을 받지 않는다 — 값이 정수임은 루프 인덱스로 보장된다.
      // COMMIT 전에 실행해야 스키마 변경과 원자적으로 묶인다(SQLite user_version 은
      // 데이터베이스 헤더에 있고 트랜잭션에 참여한다).
      db.run(`PRAGMA user_version = ${version}`);
      db.run('COMMIT');
    } catch (error) {
      db.run('ROLLBACK');
      throw error;
    }
  }
  return version;
}

File: crates/rocky-todo-core/src/migrations.rs
//! PRAGMA user_version 마이그레이션 러너 — TS 원본 `src/migrations.ts`.
//!
//! 적용 순서 = 배열 순서. 인덱스+1 이 곧 user_version. 기존 항목은 절대 수정하지 않는다.
//!
//! **규칙(SCHEMA 와의 이중 정의)**: SCHEMA(`store.rs`)는 신규 DB 를 만드는 뼈대이고 신규
//! DB 도 user_version 은 0 에서 시작한다. 어떤 컬럼을 SCHEMA 에 넣으면서 그 컬럼을 만드는
//! 마이그레이션도 배열에 있다면, 그 마이그레이션은 반드시 `PRAGMA table_info` 로 컬럼
//! 존재를 먼저 확인해야 한다 — 안 그러면 신규 DB 가 "duplicate column" 으로 기동조차
//! 못 한다.

use std::path::{Path, PathBuf};

use rusqlite::Connection;

/// 마이그레이션 하나 — 같은 트랜잭션 안에서 실행된다.
pub type MigrationFn = fn(&Connection) -> rusqlite::Result<()>;
/// 테스트 주입용 — 클로저를 받을 수 있는 동적 형태.
pub type MigrationRef<'a> = &'a (dyn Fn(&Connection) -> rusqlite::Result<()> + 'a);

fn table_columns(db: &Connection, table: &str) -> rusqlite::Result<Vec<String>> {
    let mut stmt = db.prepare(&format!("PRAGMA table_info({table})"))?;
    let names = stmt
        .query_map([], |row| row.get::<_, String>(1))?
        .collect::<rusqlite::Result<Vec<_>>>()?;
    Ok(names)
}

/// 마이그레이션 1: todo/note 에 보드별 순번(number)을 부여한다.
///
/// 기존 행에는 보드별로 created_at 순(동률이면 id 순 — 같은 밀리초 생성의 결정성)으로
/// 1부터 소급 부여한다.
fn add_numbers(db: &Connection) -> rusqlite::Result<()> {
    db.execute_batch(
        "ALTER TABLE todos ADD COLUMN number INTEGER;\n\
         ALTER TABLE notes ADD COLUMN number INTEGER;",
    )?;

    for table in ["todos", "notes"] {
        let rows: Vec<(String, Option<String>)> = {
            let mut stmt = db.prepare(&format!(
                "SELECT id, board_id FROM {table} ORDER BY board_id, created_at ASC, id ASC"
            ))?;
            let rows = stmt
                .query_map([], |row| {
                    Ok((row.get::<_, String>(0)?, row.get::<_, Option<String>>(1)?))
                })?
                .collect::<rusqlite::Result<Vec<_>>>()?;
            rows
        };
        let mut counters: std::collections::HashMap<String, i64> = std::collections::HashMap::new();
        let mut update = db.prepare(&format!("UPDATE {table} SET number = ?1 WHERE id = ?2"))?;
        for (id, board_id) in rows {
            let key = board_id.unwrap_or_default();
            let next = counters.get(&key).copied().unwrap_or(0) + 1;
            counters.insert(key, next);
            update.execute(rusqlite::params![next, id])?;
        }
    }

    db.execute_batch(
        "CREATE UNIQUE INDEX idx_todos_number ON todos(board_id, number);\n\
         CREATE UNIQUE INDEX idx_notes_number ON notes(board_id, number);\n\
         -- notes.board_id 는 nullable 이고 유니크 인덱스는 NULL 을 서로 다른 값으로 취급한다\n\
         -- — 글로벌 메모끼리의 유일성은 부분 인덱스로 따로 건다.\n\
         CREATE UNIQUE INDEX idx_notes_number_global ON notes(number) WHERE board_id IS NULL;",
    )
}

/// 마이그레이션 2: 보드에 GitHub 레포(`owner/name`)를 붙인다. SCHEMA 도 이 컬럼을
/// 만들므로 존재 확인 가드가 필수다.
fn add_board_repo(db: &Connection) -> rusqlite::Result<()> {
    if table_columns(db, "boards")?.iter().any(|c| c == "repo") {
        return Ok(());
    }
    db.execute_batch("ALTER TABLE boards ADD COLUMN repo TEXT")
}

/// 마이그레이션 3: 핸드오프 큐 테이블. 테이블은 `IF NOT EXISTS` 로 넘어가지만 인덱스는
/// 컬럼이 갖춰졌는지 확인한 뒤에만 만든다.
fn add_handoffs(db: &Connection) -> rusqlite::Result<()> {
    db.execute_batch(
        "CREATE TABLE IF NOT EXISTS handoffs (\n\
           id            TEXT PRIMARY KEY,\n\
           todo_id       TEXT NOT NULL REFERENCES todos(id),\n\
           session_id    TEXT NOT NULL,\n\
           session_name  TEXT,\n\
           session_cwd   TEXT,\n\
           note          TEXT NOT NULL DEFAULT '',\n\
           actor         TEXT NOT NULL,\n\
           status        TEXT NOT NULL CHECK (status IN ('pending','delivered','cancelled')),\n\
           created_at    TEXT NOT NULL,\n\
           delivered_at  TEXT,\n\
           delivered_via TEXT\n\
         )",
    )?;
    let columns = table_columns(db, "handoffs")?;
    let has = |name: &str| columns.iter().any(|c| c == name);
    if has("session_id") && has("status") && has("created_at") {
        db.execute_batch(
            "CREATE INDEX IF NOT EXISTS idx_handoffs_session ON handoffs(session_id, status, created_at)",
        )?;
    }
    if has("todo_id") && has("status") {
        db.execute_batch(
            "CREATE INDEX IF NOT EXISTS idx_handoffs_todo ON handoffs(todo_id, status)",
        )?;
    }
    Ok(())
}

/// 마이그레이션 4: 보드에 메인 레포 경로 — 백그라운드 세션을 띄우는 자리.
fn add_board_path(db: &Connection) -> rusqlite::Result<()> {
    if table_columns(db, "boards")?.iter().any(|c| c == "path") {
        return Ok(());
    }
    db.execute_batch("ALTER TABLE boards ADD COLUMN path TEXT")
}

/// 마이그레이션 5: 핸드오프 라이프사이클(착수/완료)과 doing 의 세션 귀속.
/// status enum 은 늘리지 않는다 — accepted/completed 는 타임스탬프로만.
fn add_handoff_lifecycle(db: &Connection) -> rusqlite::Result<()> {
    let handoff_columns = table_columns(db, "handoffs")?;
    if !handoff_columns.iter().any(|c| c == "accepted_at") {
        db.execute_batch("ALTER TABLE handoffs ADD COLUMN accepted_at TEXT")?;
    }
    if !handoff_columns.iter().any(|c| c == "completed_at") {
        db.execute_batch("ALTER TABLE handoffs ADD COLUMN completed_at TEXT")?;
    }
    if !table_columns(db, "todos")?
        .iter()
        .any(|c| c == "doing_session_id")
    {
        db.execute_batch("ALTER TABLE todos ADD COLUMN doing_session_id TEXT")?;
    }
    Ok(())
}

/// 마이그레이션 6: 보드 메타 — `description` 컬럼과 옛 key 별칭 테이블.
fn add_board_meta(db: &Connection) -> rusqlite::Result<()> {
    if !table_columns(db, "boards")?
        .iter()
        .any(|c| c == "description")
    {
        db.execute_batch("ALTER TABLE boards ADD COLUMN description TEXT")?;
    }
    db.execute_batch(
        "CREATE TABLE IF NOT EXISTS board_aliases (\n\
           key        TEXT PRIMARY KEY,\n\
           board_id   TEXT NOT NULL REFERENCES boards(id),\n\
           created_at TEXT NOT NULL\n\
         )",
    )?;
    if table_columns(db, "board_aliases")?
        .iter()
        .any(|c| c == "board_id")
    {
        db.execute_batch(
            "CREATE INDEX IF NOT EXISTS idx_board_aliases_board ON board_aliases(board_id)",
        )?;
    }
    Ok(())
}

pub const MIGRATIONS: [MigrationFn; 6] = [
    add_numbers,
    add_board_repo,
    add_handoffs,
    add_board_path,
    add_handoff_lifecycle,
    add_board_meta,
];

#[derive(Default)]
pub struct RunMigrationsOptions<'a> {
    /// 테스트에서 목록을 주입한다. 기본은 MIGRATIONS.
    pub migrations: Option<&'a [MigrationRef<'a>]>,
    /// 적용 전 DB 복사 경로. 생략하면 `${dbPath}.bak-v<현재 user_version>` 자동 계산 —
    /// 실제 시작 버전이 파일명에 남아야 여러 번 재기동을 거친 DB 의 백업 이력을 구분한다.
    pub backup_path: Option<PathBuf>,
    /// 백업 원본 경로. 없거나 todos/notes 가 비어 있으면(백업할 내용 없음) 백업 생략.
    pub db_path: Option<&'a Path>,
}

/// todos/notes 에 백업할 만한 데이터가 있는지. 판단 불가(테이블 없음 등)면 보수적으로 true.
fn has_data_worth_backing_up(db: &Connection) -> bool {
    let count = |sql: &str| -> Option<i64> { db.query_row(sql, [], |r| r.get(0)).ok() };
    match count("SELECT COUNT(*) FROM todos") {
        Some(n) if n > 0 => true,
        Some(_) => match count("SELECT COUNT(*) FROM notes") {
            Some(n) => n > 0,
            None => true,
        },
        None => true,
    }
}

/// WAL 을 메인 파일로 체크포인트한 뒤 파일 복사로 백업한다.
///
/// WAL 모드에서 메인 파일만 복사하면 마지막 체크포인트 이후 커밋이 빠진 백업이 된다.
/// 복사 실패는 경고만 남기고 진행한다(fail-open) — 백업 실패로 기동 전체가 막히면
/// "백업 없이 보드가 뜨는 것"보다 나쁘다.
fn backup_database(db: &Connection, db_path: &Path, backup_path: &Path) {
    // 체크포인트가 실패해도 최선을 다해 복사는 시도한다.
    let _ = db.execute_batch("PRAGMA wal_checkpoint(TRUNCATE)");
    if let Err(error) = std::fs::copy(db_path, backup_path) {
        eprintln!(
            "rocky-todo: pre-migration backup failed ({}) — 백업 없이 진행: {error}",
            backup_path.display()
        );
    }
}

/// user_version 보다 뒤에 있는 마이그레이션만 순서대로 적용한다.
///
/// 각 마이그레이션은 트랜잭션 안에서 실행되며, 실패하면 롤백하고 user_version 도 올리지
/// 않는다 — 다음 기동에서 재시도된다. user_version 갱신은 **같은 트랜잭션 안**에서 스키마
/// 변경과 함께 커밋된다 — COMMIT 뒤 별도로 쓰면 그 사이 프로세스가 죽었을 때 "스키마는
/// 적용, user_version 은 0" 이 남아 다음 기동에서 duplicate column 으로 영구 기동 불가.
///
/// 적용 후 최종 user_version 을 돌려준다.
pub fn run_migrations(db: &Connection, options: RunMigrationsOptions) -> rusqlite::Result<i64> {
    let default_refs: Vec<MigrationRef> = MIGRATIONS.iter().map(|f| f as MigrationRef).collect();
    let migrations: &[MigrationRef] = options.migrations.unwrap_or(&default_refs);
    let current: i64 = db.query_row("PRAGMA user_version", [], |r| r.get(0))?;
    if current >= migrations.len() as i64 {
        return Ok(current);
    }

    if let Some(db_path) = options.db_path {
        if db_path.exists() && has_data_worth_backing_up(db) {
            let backup_path = options
                .backup_path
                .unwrap_or_else(|| PathBuf::from(format!("{}.bak-v{current}", db_path.display())));
            backup_database(db, db_path, &backup_path);
        }
    }

    let mut version = current;
    for (i, migration) in migrations.iter().enumerate() {
        if (i as i64) < current {
            continue;
        }
        db.execute_batch("BEGIN")?;
        let applied = migration(db).and_then(|()| {
            version = i as i64 + 1;
            // PRAGMA 는 바인딩을 받지 않는다 — 값이 정수임은 루프 인덱스로 보장된다.
            // COMMIT 전에 실행해야 스키마 변경과 원자적으로 묶인다.
            db.execute_batch(&format!("PRAGMA user_version = {version}"))?;
            db.execute_batch("COMMIT")
        });
        if let Err(error) = applied {
            let _ = db.execute_batch("ROLLBACK");
            return Err(error);
        }
    }
    Ok(version)
}
